crates/starknet/src/services/add_declare_transaction.rs

```diff
@@ -0,0 +1,225 @@
+use starknet_in_rust::transaction::Declare;
+use starknet_types::{
+    error::Error,
+    felt::{ClassHash, TransactionHash},
+    traits::HashProducer,
+    DevnetResult,
+};
+
+use crate::{
+    traits::StateChanger,
+    transactions::{declare_transaction::DeclareTransactionV1, StarknetTransaction, Transaction},
+    Starknet,
+};
+
+impl Starknet {
+    pub fn add_declare_transaction_v1(
+        &mut self,
+        declare_transaction: DeclareTransactionV1,
+    ) -> DevnetResult<(TransactionHash, ClassHash)> {
+        let mut declare_transaction = declare_transaction;
+
+        let class_hash = declare_transaction.contract_class.generate_hash()?;
+        let transaction_hash = declare_transaction.generate_hash()?;
+        declare_transaction.transaction_hash = Some(transaction_hash);
+
+        let transaction = Declare {
+            class_hash: class_hash.into(),
+            sender_address: declare_transaction.sender_address.try_into()?,
+            tx_type: starknet_in_rust::definitions::transaction_type::TransactionType::Declare,
+            validate_entry_point_selector:
+                starknet_in_rust::definitions::constants::VALIDATE_DECLARE_ENTRY_POINT_SELECTOR
+                    .clone(),
+            version: declare_transaction.version().into(),
+            max_fee: declare_transaction.max_fee,
+            signature: declare_transaction.signature.iter().map(|felt| felt.into()).collect(),
+            nonce: declare_transaction.nonce.into(),
+            hash_value: transaction_hash.into(),
+            contract_class: declare_transaction.contract_class.clone().try_into()?,
+        };
+
+        transaction.verify_version()?;
+
+        if transaction.max_fee == 0 {
+            return Err(Error::StarknetInRustTransactionError(
```

Comment:
Why `StarknetInRustTransactionError`? How does this look for the end user? Does it somewhere get wrapped as a JSON-RPC error? Where? Or is this addressed in https://github.com/0xSpaceShard/starknet-devnet-rs/pull/8?

Reply:
Errors of the core project are transformed by whatever endpoint is consuming it

Reply:
Where does this happen? Can you point me, please?

Reply:
In endpoints.rs when the core methods are invoked, they will return either an Ok or Err. Then this error will be mapped to an ApiError

Reply:
Ah so that's in #8, we don't have to discuss that here. Resolving.